### DISCUSSION TITLE
Use CAdES with Signer.async_sign_raw

### DISCUSSION THREAD

**Discussion by ruimelodev** [MatthiasValvekens/pyHanko#339] [Q&A] [answered]
Hi @MatthiasValvekens,

The esign service I use to sign my PDF documents recently did some changes to be compliant with the EU commission regarding electronic signatures.
They requested me to update my code to use PAdES-BASELINE-B signature type and they gave me the following example using the iText library:
```java
PdfSigner pdfSigner = new PdfSigner(pdfReader,
    new FileStream(DestinationSignedPdf, FileMode.OpenOrCreate), stampingProperties);

pdfSigner.GetSignatureAppearance().SetPageNumber(1).SetReason(Reason).SetLocation(Location);
pdfSigner.SetFieldName(SignatureFieldName);

IExternalSignature pks = new ServerSignature(Iteration, SourcePlainPdf, AuthenticationToken, QscdServerBaseAddress, CertificateAlias, Seed, Secret, HashAlgorithmOID);

// Change from this:
// pdfSigner.SignDetached(pks, certificates.ToArray(), null, null, null, 0, PdfSigner.CryptoStandard.CMS); // PKCS7-B
// To this:
pdfSigner.SignDetached(pks, certificates.ToArray(), null, null, null, 0, PdfSigner.CryptoStandard.CADES); // PAdES-BASELINE-B
```

I noticed the Signer class has some signing methods where I can just pass `use_cades=True`, but I need to use the method `async_sign_raw` because I'm bound to this external service.

I suppose I need to set a `CAdESSignedAttributeProviderSpec` but I'm a bit lost on how to do it within the `async_sign_raw` method.

This is my code so far:
```python
class ExternalServiceSigner(Signer):
    prefer_pss = False
    embed_roots = True
        
    async def async_sign_raw(self, data: bytes, digest_algorithm: str, dry_run=False) -> bytes:
        if dry_run:
            return bytes(256)
        
        hash_spec = get_pyca_cryptography_hash(self.signature_mechanism.hash_algo)
        md = hashes.Hash(hash_spec)
        md.update(data)
        
        # External service request
        response = request_external_service_to_sign_hash(
            b64encode(md.finalize()).decode("utf-8"))
        
        return b64decode(response.json()[0]["hashSig"])


my_signer = ExternalServiceSigner(
    signing_cert=cert, 
    signature_mechanism=algos.SignedDigestAlgorithm(
        {'algorithm': 'sha256_rsa'}
    ))
meta = signers.PdfSignatureMetadata(field_name='sig')
output = signers.sign_pdf(w, meta, signer=my_signer, output=None)

with open(f'signed-file-{time.time()}.pdf', 'wb') as outf:
    outf.write(output.getvalue())
```

Thank you for your help.

**Comment by MatthiasValvekens** [comment, marked as answer]
Hi @ruimelodev, 

You're in luck: with that calling convention you can just pass `use_pades=True` (EDIT: sorry, `subfilter=PADES` is the current syntax) in `PdfSignatureMetadata` and pyHanko will do the rest. No need to change your custom signer, but you might need some more config if you also want timestamps to be included (depends on the specific PAdES level you need).

The CAdES signed attributes thing is for adding optional attributes that are defined in CAdES but not in "core" CMS. They're not needed for baseline compliance, regardless of the level.

**Reply by ruimelodev** [reply]
Thank you for your fast response, adding `subfilter=SigSeedSubFilter.PADES` did the trick. 
However, passing the output file in a validator returns this error:
![image](https://github.com/MatthiasValvekens/pyHanko/assets/16306125/df6ee4b2-487f-4ffc-ac51-62af99a5e6c2)
Am I missing any additional config?
This seems to be an issue with the external service I'm using but I just wanted to confirm with you first.

**Reply by MatthiasValvekens** [reply]
Are you embedding all relevant certs, including the signer's certificate and all applicable intermediate CA certs? Or perhaps signing with a testing key?

Edit: in your snippet, I don't see any code to embed intermediate CA certs, FWIW. I'm on mobile right now, but the Signer base class API docs should be of help.

**Reply by ruimelodev** [reply]
I think that might be the issue. 

My certificate file has 3 certificates in it, 1 for my organization and other 2 from the CA.
I've checked the Signer class and I believe I need to pass the CA certs as `attribute_certs`, however, I'm having some trouble using the class `AttributeCertificateV2`, I'm not sure how to pass the cert data into it and I was unable to find examples in the docs.

This is what I have so far:
```python
def resolve_cert_file():
    sign_cert = None
    attr_certs = []

    data = b64decode(fetch_cert()["cert_64"])
    ci = cms.ContentInfo.load(data)

    for cert_choice in ci['content']['certificates']:
        cert = cert_choice.chosen
        if cert.subject.native['organization_name'] == "My organization name":
            sign_cert = cert
        else:
            attr_certs.append(cms.AttributeCertificateV2(...))
    return sign_cert, attr_certs
```

**Reply by MatthiasValvekens** [reply]
No, you don't need `attribute_certs` for this one :). Attribute certificates are a fairly advanced thing that is not all that commonly used outside of specialised workflows.

You'll want to use the `cert_registry` argument to `Signer.__init__(...)`, like so:

```
        cert_registry=SimpleCertificateStore.from_certs(
            list(load_certs_from_pemder('path/to/ca/certificate/file'))
        )
```

I'm assuming that the file containing your CA certs is in PEM format here. If not, the loading function might look slightly different, but that's the gist of it.
